# IC Data Wrangling and Subsetting Exercise

## List of Contents:
## 01. Importing Libraries
## 02. Importing Data
## 03. Wrangling Procedures
## 04. Transposing Data
## 05. Data Dictionaries
## 06. Subsetting
## 07. Wrangling Procedures Continued
## 08. Exporting Data

## --


## 01. Importing Libraries

In [1]:
#Importing libraries
import pandas as pd
import numpy as np
import os

## 02. Importing Data

In [2]:
#Assigning path for import
path = r'C:\Users\Mosspants\Desktop\CareerFoundry\A4_Python_Fundamentals_for_Data_Analysts\Instacart Basket Analysis_10_2023'

In [3]:
#Importing orders.csv
df_ords = pd.read_csv(os.path.join(path, '02_Data', 'Original_Data', 'orders.csv'), index_col = False)

In [4]:
#Importing products.csv
df_prods = pd.read_csv(os.path.join(path, '02_Data', 'Original_Data', 'products.csv'), index_col = False)

In [5]:
#Importing departments.csv
df_dep = pd.read_csv(os.path.join(path, '02_Data', 'Original_Data', 'departments.csv'), index_col = False)

## 03. Wrangling Procedures

### Dropping columns

In [6]:
#Removing eval_set from orders by making a new dataframe
df_ords_2 = df_ords.drop(columns = ['eval_set'])

In [7]:
#Checking output
df_ords_2.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

### Checking for missing values

In [8]:
#Value counts on days_since_prior_order column in orders data to look for NaN values
df_ords_2['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

### Renaming columns

In [9]:
#Renaming order_dow column to order_day_of_week in orders
df_ords_2.rename(columns = {'order_dow' : 'order_day_of_week'}, inplace = True)

In [10]:
#Checking output
df_ords_2.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

### Changing a variable's data type

In [11]:
#Changing data type of 'order_id' in orders
df_ords_2['order_id'] = df_ords_2['order_id'].astype('str')

In [12]:
#Checking output
df_ords_2['order_id'].dtype

dtype('O')

## 04. Transposing Data

In [13]:
#Transposing department data
df_dep_t = df_dep.T

In [14]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [15]:
#Resetting index of transposed department dataframe
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [16]:
#Creating new header for departments
new_header = df_dep_t.iloc[0]

In [17]:
#Removing first row in departments
df_dep_t_new = df_dep_t[1:]

In [18]:
#Adding new header to departments
df_dep_t_new.columns = new_header

In [19]:
#Checking output
df_dep_t_new.head()

department_id department
1                 frozen
2                  other
3                 bakery
4                produce
5                alcohol

## 05. Data Dictionaries

In [20]:
#Creating data dictionary of departments
data_dict = df_dep_t_new.to_dict('index')

In [21]:
#Checking output
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [22]:
#Calling '19' from departments data dictionary
print(data_dict.get('19'))

{'department': 'snacks'}


## 06. Subsetting

In [23]:
#Creating a subset of products data with department_id of 19
df_snacks = df_prods[df_prods['department_id'] == 19]

In [24]:
#Checking subset
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

## 07. Wrangling Procedures Continued

### Change another variable in df_ords that doesn't need to be included as a numeric variable and changed to a sutible format.

In [25]:
#changing user_id in orders to string
df_ords_2['user_id'] = df_ords_2['user_id'].astype('str')

In [26]:
#Checking change
df_ords_2['user_id'].dtype

dtype('O')

### Find another variable in df_ords and change name without overwriting the dataframe.

In [27]:
#Renaming 'days_since_prior_order' to 'days_count_from_last_order' in orders without overwriting the dataframe
df_ords_2.rename(columns = {'days_since_prior_order' : 'days_count_from_last_order'})

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0        2539329       1             1                  2                  8   
1        2398795       1             2                  3                  7   
2         473747       1             3                  3                 12   
3        2254736       1             4                  4                  7   
4         431534       1             5                  4                 15   
...          ...     ...           ...                ...                ...   
3421078  2266710  206209            10                  5                 18   
3421079  1854736  206209            11                  4                 10   
3421080   626363  206209            12                  1                 12   
3421081  2977660  206209            13                  1                 12   
3421082   272231  206209            14                  6                 14   

         days_count_from_last_order  
0                               NaN  
1                              15.0  
2                              21.0  
3                              29.0  
4                              28.0  
...                             ...  
3421078                        29.0  
3421079                        30.0  
3421080                        18.0  
3421081                         7.0  
3421082                        30.0  

[3421083 rows x 6 columns]

In [28]:
#Check to see if new name overwrote original column
df_ords_2.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0  2539329       1             1                  2                  8   
1  2398795       1             2                  3                  7   
2   473747       1             3                  3                 12   
3  2254736       1             4                  4                  7   
4   431534       1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

### Busiest hour for placing orders.

In [29]:
#Frequency count for 'order_hour_of_day' in orders
df_ords_2['order_hour_of_day'].value_counts()

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

### Hour 10 (10:00AM) is the busiest hour for placing orders.

### Determine meaning behind the value 4 in "department_id" in df_prods using a data dictionary.

In [30]:
#Calling '4' from departments data dictionary
print(data_dict.get('4'))

{'department': 'produce'}


### 4 represents produce in the department_id.

### Create a subset containing breakfast item sales.

In [31]:
#Based on the departments data dictionary, breakfast items are department_id 14. Printing department_id for verification
print(data_dict.get('14'))

{'department': 'breakfast'}


In [32]:
#Creating breakfast (df_breakfast) subset from departments
df_breakfast = df_prods[df_prods['department_id'] == 14]

In [33]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

### Create a subset of dinner party items including alcohol, deli, beverages and meat/seafood.

In [34]:
#Verifying department_id numbers for alcohol, deli, beverages, and meat/seafood from departments data dictionary
print(data_dict.get('5'))
print(data_dict.get('20'))
print(data_dict.get('7'))
print(data_dict.get('12'))

{'department': 'alcohol'}
{'department': 'deli'}
{'department': 'beverages'}
{'department': 'meat seafood'}


In [35]:
#Creating dinner party items (df_dinner) subset from departments
df_dinner = df_prods.loc[df_prods['department_id'].isin([5, 20, 7, 12])]

In [36]:
df_dinner

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

### How many rows does the last dataframe created have?

In [37]:
#Above output of dinner subset shows 7,650 rows, but can also use .shape to find the information
df_dinner.shape

(7650, 5)

### The dinner subset has 7,650 rows.

### All information about 'user_id' of '1'.

In [47]:
#Creating subset of user 1 from orders 
df_user_1 = df_ords_2[df_ords_2['user_id'] == '1'] 

In [48]:
df_user_1

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   
5   3367565       1             6                  2                  7   
6    550135       1             7                  1                  9   
7   3108588       1             8                  1                 14   
8   2295261       1             9                  1                 16   
9   2550362       1            10                  4                  8   
10  1187899       1            11                  4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

In [49]:
#Checking shape of user 1 subset
df_user_1.shape

(11, 6)

### Basic statistics based on 'user_id' of '1'.

In [50]:
#Using describe() to get descriptive statistics on user 1 subset
df_user_1.describe()

order_number  order_day_of_week  order_hour_of_day  \
count     11.000000          11.000000          11.000000   
mean       6.000000           2.636364          10.090909   
std        3.316625           1.286291           3.477198   
min        1.000000           1.000000           7.000000   
25%        3.500000           1.500000           7.500000   
50%        6.000000           3.000000           8.000000   
75%        8.500000           4.000000          13.000000   
max       11.000000           4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

### user_id 1 has 11 orders. Average hour of day is at 10:00AM. 
### Frequently places orders on the 4th day of the week (Wednesday for Sunday start week, Thursday for Monday start week.)
### Average time between orders is 19 days. Placed order 2295261 two hours after 3108588 on the same day. Max time between orders is 30 days.

## 08. Exporting Data

### Exporting df_ords as "orders_wrangled.csv"

In [51]:
df_ords_2.to_csv(os.path.join(path, '02_Data', 'Prepared_Data', 'orders_wrangled.csv'))

### Exporting df_dept_t_new as "department_wrangled.csv"

In [52]:
df_dep_t_new.to_csv(os.path.join(path, '02_Data', 'Prepared_Data', 'departments_wrangled.csv'))